### ISSUE TITLE
Add typing to RangeSet to avoid typing errors when iterating

### ISSUE THREAD

**Issue by carl-aa** [Pyomo/pyomo#2078]
## Summary

Currently you are able to iterate over `RangeSet` instances in a `ConcreteModel` (e.g. `for x in y` where y is a RangeSet instance), but I get typing warnings saying ```Expected type `collections.iterable`, got `RangeSet` instead```. It would be helpful to expose type hints for `RangeSet` so that these type warnings could be avoided.

### Rationale

It would make development in type aware environments like PyCharm more ergonomic.

### Description

I'm not exactly sure what's missing to make the iteration error go away. I would need to research typing issues more.

### Additional information

The `Set` instances work just fine, so it may be easy to compare the two classes to see what change is needed.

**Comment by jsiirola** [comment]
I think I know what is going on, but I am not convinced that there is an easy fix (nor that the fix would be "correct").

The challenge is that not all `RangeSet` instances are iterable.

`RangeSet()` is a factory that actually creates either a `FiniteScalarRangeSet` (e.g., from `RangeSet(5)`) or an `InfiniteScalarRangeSet` (e.g., from `RangeSet(1, 5, 0)`, or the `Reals` set -- which is a special instance of `RangeSet(ranges=(NumericRange(None, None, 0),))`).  

Non-finite sets are (obviously) not iterable.

`Set` works out because of a bit of a fluke: `Set` inherits from `IndexedComponent` (`RangeSet` does not support indexing, and therefore only inherits from `Component`).  The type hints don't complain about `Set` because they see the `IndexedComponent.__iter__` method (which is defined for iterating over the index that defines the `IndexedComponent` -- borrowing from the `MutableMapping` model).  This isn't a huge deal, as all "real" `Set` objects (e.g., `OrderedScalarSet`, `SortedScalarSet`, `FiniteScalarSet`, etc) are all finite sets (and will support iteration) and override the `__iter__` method.

**Comment by carl-aa** [comment]
Ahh, I did not know about the infinite case. In my particular case, I simply switched over to using `Set`, since it didn't seem to matter for my modeling use case.

**Comment by jsiirola** [comment]
Non-finite range sets are a pretty niche topic (they are mostly used for defining domains).  A reasonable solution might be to break away from the Factory pattern and force users to explicitly create `InfiniteRangeSet` objects.  That way `RangeSet` could become an alias for `FiniteRangeSet`, and type hints could make sense.

...there might be some issues around ctypes, but that should be resolvable.  It actually may help motivate something I have wanted to do for a while now (switch `RangeSet`'s `ctype` over to `Set`).  I am interested in hearing others' thoughts on the issue...
